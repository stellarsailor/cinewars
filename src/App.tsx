import { useEffect } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom'

import useList from './hooks/useList';
import Character from './pages/Character';
import Main from './pages/Main';
import AppLayout from './components/AppLayout';
import Loader from './components/Loader';

function App() {
  
  const { status, initialFetch } = useList();

  useEffect(() => {
    initialFetch(1)
  },[initialFetch])

  return (
    <Router>
      <Switch>
        {status === 'pending' && <Loader global />}
        <AppLayout>
          <Route path="/" exact>
            <Main />
          </Route>
          <Route path="/character/:id">
            <Character />
          </Route>
        </AppLayout>
      </Switch>
    </Router>
  );
}

export default App;
