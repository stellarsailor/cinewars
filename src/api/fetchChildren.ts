import { CharacterProfile } from '../types/CharacterProfile';

export async function fetchChildren(data: CharacterProfile) {
  // in iOS safari, cors issue occurs in localhost. To prevent it, replace http url to https.
  // there was cors issue, but it solves after adding meta tag <meta http-equiv="Content-Security-Policy" content="upgrade-insecure-requests">
  // it was due to SWAPI server is being served on http
  const res = await fetch(data.homeworld.replace('http:', 'https:'));
  const homeworldData = await res.json();

  const filmsPromisesArray = data.films?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  // ! non-null assertion operator for -> Argument of type 'Promise<any>[] | undefined' is not assignable to parameter of type 'Iterable<any>'.
  const filmsData = await Promise.all(filmsPromisesArray!);

  const speciesPromisesArray = data.species?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const speciesData = await Promise.all(speciesPromisesArray!);

  const vehiclesPromisesArray = data.vehicles?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const vehiclesData = await Promise.all(vehiclesPromisesArray!);

  const starshipsPromisesArray = data.starships?.map(async (url: string) =>
    fetch(url.replace('http:', 'https:')).then((res) => res.json())
  );
  const starshipsData = await Promise.all(starshipsPromisesArray!);

  return {
    ...data,
    homeworldData: homeworldData,
    filmsData: filmsData,
    speciesData: speciesData,
    vehiclesData: vehiclesData,
    starshipsData: starshipsData,
  };
}
