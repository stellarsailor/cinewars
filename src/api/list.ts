export async function getList(page: number) {
  let url = `http://swapi.dev/api/people/?page=${page}`;
  try {
    const res = await fetch(url)
    const data = await res.json()
    return data
  } catch (err) {
    console.log(err)
  }
}