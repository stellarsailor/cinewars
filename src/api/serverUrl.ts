export const serverUrl = 'https://swapi.dev/api';
