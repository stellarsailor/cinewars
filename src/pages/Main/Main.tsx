import { useEffect, useState } from "react"
import { Link } from "react-router-dom"
import useList from "../../hooks/useList"

export type MainProps = {}

function Main({}: MainProps) {

  const { list, onSetPage, onFetchListBy } = useList();
  
  useEffect(() => {
    onFetchListBy(list.page)
  },[list.page, onFetchListBy])

  if(list.status === 'loading') {
    return <div>loading..</div>
  }

  return (
    <div>
      {list.characters && list.characters.length !== 0 && list.characters.map( (v: any, index: number) => (
        <Link to={`/character/${parseInt(v.url.slice(28))}`} key={v.url}>
          <div>
            {v.name}
          </div>
        </Link>
      ))}
      <div>
        <div onClick={() => onSetPage(list.page - 1)}>prev</div>
        {list.page}
        <div onClick={() => onSetPage(list.page + 1)}>next</div>
      </div>
    </div>
  )
}

export default Main