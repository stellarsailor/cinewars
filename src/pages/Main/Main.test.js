import { 
  render, 
  fireEvent, 
  waitFor 
} from "@testing-library/react";
import { 
  createStore, 
  applyMiddleware 
} from 'redux'
import { Provider } from 'react-redux'
import Thunk from 'redux-thunk';
import rootReducer from '../../modules';
import App from '../../App'
import Main from "./Main";
import { serverUrl } from "../../api/serverUrl";
import axios from 'axios';
// to resolve cors problem (https://github.com/axios/axios/issues/1754)
axios.defaults.adapter = require('axios/lib/adapters/http');

const store = createStore(rootReducer, applyMiddleware(Thunk))

describe('<Main />', () => {
  it('matches snapshot', () => {
    const utils = render(
      <Provider store={store}>
        <Main />
      </Provider>
    );
    expect(utils.container).toMatchSnapshot();
  });

  it('shows the table properly', () => {
    const utils = render(
      <Provider store={store}>
        <Main />
      </Provider>
    );
    utils.getByText('Name')
    utils.getByText('Birth')
    utils.getByText('Height')
    utils.getByText('Mass')

    utils.getByText('< Prev')
    utils.getByText('1')
    utils.getByText('Next >')

    utils.getByAltText("stormtrooper's portrait")
  });

  it("shows character list properly", async () => {
    const utils = render(
      <Provider store={store}>
        <App>
          <Main />
        </App>
      </Provider>
    );
    let url = `${serverUrl}/people`
    const res = await axios(url)
    const data = res.data

    utils.getByText(data.results[0].name)
    utils.getAllByText(data.results[0].birth_year)
    utils.getAllByText(data.results[0].height)
    utils.getAllByText(data.results[0].mass)
  })
  
  it('should remain page 1 when press Prev at page 1 and shows next page when click Next >', async () => {
    const utils = render(
      <Provider store={store}>
        <Main />
      </Provider>
    );
    const prevPageButton = utils.getByText('< Prev')
    const pageNumber = utils.getByText('1')
    const nextPageButton = utils.getByText('Next >')

    fireEvent.click(prevPageButton)

    expect(pageNumber).toHaveTextContent('1')

    fireEvent.click(nextPageButton)

    let url = `${serverUrl}/people/?page=2`
    const res = await axios(url)
    const data = res.data

    utils.getByText(data.results[0].name)
    utils.getAllByText(data.results[0].birth_year)
    utils.getAllByText(data.results[0].height)
    utils.getAllByText(data.results[0].mass)
  });

});