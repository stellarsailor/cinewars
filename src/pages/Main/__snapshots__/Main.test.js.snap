// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Main /> matches snapshot 1`] = `
<div>
  <div
    class="sc-iwyYcG diVfFX"
  >
    <div
      class="sc-idOhPF lbVZAh"
    >
      <div
        style="margin-left: 0px; margin-right: 0px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: center;"
      >
        <div
          style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 0px; padding-right: 0px; width: 66.66666666666667%; flex-basis: 66.66666666666667%; flex-grow: 0; flex-shrink: 0; max-width: 66.66666666666667%; margin-left: 0%;"
        >
          <div
            class="sc-dIUggk jUgLus"
            style="opacity: 0; transform: translateY(20px) translateZ(0);"
          >
            <div
              class="sc-hHftDr ePQka-d"
            >
              <table>
                <thead>
                  <tr>
                    <th
                      class="sc-bBXqnf kKsWNn"
                    >
                      Name
                    </th>
                    <th
                      class="sc-bBXqnf kKsWNn"
                    >
                      Birth
                    </th>
                    <th
                      class="sc-bBXqnf kKsWNn"
                    >
                      Height
                    </th>
                    <th
                      class="sc-bBXqnf kKsWNn"
                    >
                      Mass
                    </th>
                  </tr>
                </thead>
                <tbody />
              </table>
              <div
                class="sc-dmlrTW csAspt"
              >
                <span
                  class="sc-fKFyDc hlgfOM"
                  disabled=""
                >
                  &lt;
                   Prev
                </span>
                <span
                  class="sc-kfzAmx jIEcjh"
                >
                  1
                </span>
                <span
                  class="sc-fKFyDc hlgfOM"
                  disabled=""
                >
                  Next 
                  &gt;
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-kEjbxe khGPtF"
    >
      <div
        style="margin-left: 0px; margin-right: 0px; display: flex; flex-wrap: wrap; flex-grow: 0; flex-shrink: 0; align-items: normal; justify-content: center;"
      >
        <div
          style="box-sizing: border-box; min-height: 1px; position: relative; padding-left: 0px; padding-right: 0px; width: 66.66666666666667%; flex-basis: 66.66666666666667%; flex-grow: 0; flex-shrink: 0; max-width: 66.66666666666667%; margin-left: 0%;"
        >
          <div
            class="sc-iqHYGH ihcdgJ"
            style="opacity: 0; transform: translateY(20px) translateZ(0);"
          >
            <div
              class="sc-dQppl eHjXOW"
            >
              <div
                class="sc-crrsfI bYnux"
              >
                <img
                  alt="stormtrooper's portrait"
                  src="/images/stormtrooper.png"
                  style="width: auto; height: 100%; position: relative;"
                />
              </div>
              <div
                class="sc-bqyKva kmvbgE"
              >
                <div
                  class="Typist "
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
